src/browser/application.ts

```diff
@@ -0,0 +1,111 @@
+import { Logger } from "pino";
+import { v4 as uuid } from "uuid";
+import { IApplication } from "../base/application";
+import { IInputHandler, IInputMessage } from "../base/input-handler";
+import { ISignalProvider } from "../base/signal-provider";
+import { IStreamProvider } from "../base/stream-provider";
+import { IWebrtcProvider } from "../base/webrtc-provider";
+
+/**
+ * Application constructor options
+ */
+export interface IApplicationOpts {
+  /**
+   * A logger
+   */
+  logger: Logger;
+
+  /**
+   * The name of the capture window
+   */
+  captureWindowTitle: string;
+
+  /**
+   * The ice servers to use
+   */
+  iceServers: RTCIceServer[];
+
+  /**
+   * The signal provider to use
+   */
+  signalProvider: ISignalProvider;
+
+  /**
+   * The stream provider to use
+   */
+  streamProvider: IStreamProvider;
+
+  /**
+   * The webrtc provider to use
+   */
+  webrtcProvider: IWebrtcProvider;
+
+  /**
+   * The input handler to use
+   */
+  inputHandler: IInputHandler;
+}
+
+/**
+ * A browser application - orchestrates the streamer experience
+ */
+export class Application implements IApplication {
+  private opts: IApplicationOpts;
+
+  /**
+   * Default ctor
+   * @param opts ctor options
+   */
+  constructor(opts: IApplicationOpts) {
+    this.opts = opts;
+  }
+
+  /**
+   * Internal boot helper
+   */
+  public async boot() {
+    const {
+      logger,
+      captureWindowTitle,
+      iceServers,
+      signalProvider,
+      streamProvider,
+      webrtcProvider,
+      inputHandler } = this.opts;
+    logger.info("Browser: initializing application");
+
+    const devices = await streamProvider.enumerateDevices((e) => e.name === captureWindowTitle);
+    const selectedDevice = devices[0];
+    const stream = await streamProvider.createStream(selectedDevice);
+
+    webrtcProvider.initialize(iceServers, stream);
+    await signalProvider.signIn(`${captureWindowTitle}.${uuid()}`);
+
+    let remotePeerId: string;
```

Comment:
note: throughout this refactor, `peerId` is a string - to allow us to support signal implementations that use non-number uuids. Since we never interact with `peerId` as a number anyway, this works just great. A minor change, to help with future work.